src/components/Typography/index.tsx

```diff
@@ -0,0 +1,346 @@
+    ${({ maxLines }) => (maxLines ? styles.clamp({ maxLines }) : undefined)}
+`
+
+export const typographyVariants = Object.keys(variantTags)
+
+type TypographyProps = {
+  align?: string
+  as?: string
+  children: React.ReactNode
+  color?: string
+  ellipsis?: boolean
+  fontWeight?: number
+  lineHeight?: string | number
+  maxLines?: number
+  onFocus?: (event: FocusEvent) => void
+  onMouseEnter?: (event: MouseEvent) => void
+  tooltipProps?: Partial<TooltipProps>
+  variant?: TypographyVariant
+  width?: string | number
+}
+
+const Text = forwardRef(
+  (
+    {
+      onMouseEnter,
+      onFocus,
+      tooltipProps = {},
+      variant = 'default',
+      children,
+      as,
+      color,
+      align,
+      lineHeight,
+      maxLines = 0,
+      ellipsis = false,
+      fontWeight,
+      width,
+      ...props
+    }: TypographyProps,
+    ref,
+  ) => (
+    <StyledText
+      ref={ref}
+      onMouseEnter={onMouseEnter}
+      onFocus={onFocus}
+      {...tooltipProps}
+      {...props}
+      width={width}
+      as={as || (variant ? variantTags[variant] : undefined)}
+      color={color}
+      variant={variant}
+      align={align}
+      ellipsis={ellipsis}
+      lineHeight={lineHeight}
+      fontWeight={fontWeight}
+      maxLines={maxLines}
+    >
+      {children}
+    </StyledText>
+  ),
+)
+
+const TextWithTooltip = ({ children, ...props }: TypographyProps) => {
+  const isTruncated = useCallback((target = {}) => {
+    // If the text is really truncated
+    const { offsetWidth, scrollWidth } = target as {
+      offsetWidth: number
+      scrollWidth: number
+    }
+
+    return offsetWidth < scrollWidth
+  }, [])
+
+  const finalStringChildren = recursivelyGetChildrenString(children)
+
+  return (
+    <Tooltip text={finalStringChildren}>
+      {({ onMouseEnter = () => {}, onFocus = () => {}, ...tooltipProps }) => (
+        <Text
+          {...props}
+          onMouseEnter={(ev: MouseEvent) => {
+            if (isTruncated(ev.currentTarget)) {
+              onMouseEnter()
+            }
+          }}
+          onFocus={(ev: FocusEvent) => {
+            if (isTruncated(ev.currentTarget)) {
+              onFocus()
+            }
+          }}
+          tooltipProps={tooltipProps}
+        >
+          {children}
+        </Text>
+      )}
+    </Tooltip>
+  )
+}
+
+const Typography = forwardRef((props: TypographyProps, ref) => {
+  const Component = props.ellipsis ? TextWithTooltip : Text
+
+  return <Component ref={ref} {...props} />
+})
+
+Typography.propTypes = {
+  align: PropTypes.string,
+  as: PropTypes.string,
+  children: PropTypes.node.isRequired,
+  color: PropTypes.string,
+  ellipsis: PropTypes.bool,
+  fontWeight: PropTypes.number,
+  lineHeight: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
+  maxLines: PropTypes.number,
+  onFocus: PropTypes.func,
+  onMouseEnter: PropTypes.func,
+  tooltipProps: PropTypes.shape({}),
+  variant: PropTypes.oneOf(typographyVariants),
```

Comment:
There is `width?: string | number` for type definition but the is none in propTypes